rfd/0231-in-band-mfa-ssh-sessions.md

```diff
@@ -0,0 +1,548 @@
+issues by:
+
+1. MFA enforcement is tightly integrated with session creation (i.e., in-band), ensuring that authentication factors are
+   directly bound to each session and mitigating the risk of bypasses like those seen in
+   [CVE-2025-49825](https://github.com/gravitational/Teleport/security/advisories/GHSA-8cqv-pj7f-pwpc).
+1. MFA complexity is moved from clients to the SSH service, so clients only need to connect to the SSH service and
+   resolve an MFA challenge only if required, streamlining the session establishment process and reducing the risk of
+   implementation errors.
+1. Per-session MFA certificates can be completely removed, eliminating a single credential representing multiple factors
+   of authentication.
+
+In summary, this RFD proposes a more secure and streamlined approach to MFA enforcement for SSH sessions by integrating
+MFA checks directly into the session establishment process, reducing client complexity, and eliminating the need for
+per-session MFA certificates.
+
+## Non-Goals
+
+1. This RFD does not propose changes to other Teleport access protocols such as Kubernetes, databases, desktops, etc.
+   However, the architecture could be extended to these protocols in the future.
+
+## Details
+
+### UX
+
+No changes are expected since this is an internal change.
+
+### High-Level Flow
+
+The client will first dial its target SSH host using the Proxy and its standard Teleport client certificate. The Proxy
+will then resolve the target cluster and host, and invoke the `EvaluateSSHAccess` RPC of the [Decision
+service](https://github.com/gravitational/Teleport.e/blob/master/rfd/0024e-access-control-decision-api.md). Upon
+receiving a permit from the Decision service, the Proxy will staple the permit to the connection and open a connection
+to the respective cluster's SSH service.
+
+The SSH service will authenticate the client using the provided client certificate and the stapled permit. The SSH
+service will then check if MFA is required for the session by examining the permit.
+
+If MFA is _not required_, the SSH service will then proceed to establish the SSH session.
+
+If MFA _is required_, the SSH service will send a Protobuf [`MFAPrompt` message](#ssh-keyboard-interactive-messages) via
+the SSH [keyboard-interactive channel](https://www.rfc-editor.org/rfc/rfc4256) to inform the client that MFA is needed.
+
+The client must then invoke the `CreateChallenge` RPC of the MFA service, providing a _session identifying payload
+(SIP)_ along with any relevant request metadata. The MFA service will respond to the client with a challenge that must
+be solved. The client must solve the MFA challenge and send a Protobuf
+[`MFAPromptAnswer`](#ssh-keyboard-interactive-messages) message back to the SSH service via the keyboard-interactive
+channel.
+
+Once the MFA challenge response is received from the client, the SSH service will invoke the `ValidateChallenge` RPC
```

Comment:
How will this work in trusted clusters?  SSH services in leaf clusters do not have access to the root cluster's API, but IIUC MFA ceremonies are always performed against the root auth.

I haven't been following the progression of this feature in detail, but didn't an earlier iteration have challenge-validation happening on the proxy, or am I misremembering that?  Leaf proxies _do_ have access to the root auth server API, so that might be a point in favor of proxy-based validation if y'all don't already have a plan for trusted clusters.

Reply:
I think there's customers that currently rely on the ability to do direct connections even in leaf clusters, so that might not be a complete solution anyway. If we're willing to drop support for that it'd be nice to actually terminate all client connections from a parent cluster at the proxy so peripheral agents won't need to care about nonlocal trust anymore.

Reply:
The leaf proxies don't currently access the root API, do they? They _could_ because the RemoteProxy builtin role is actually symmetric, but all current API access is done by proxies in the root to the API of the leaf and I'm not sure if there's a sanctioned way for the leaf to connect to the root API.

Reply:
Direct communication between clusters is one way, from Root to Leaf.